File: v0.2/golang/scenario/guardrail.go
// SPDX-License-Identifier: Apache-2.0
//
// Based on the Provenance Identity Continuity (PIC) Model created by Nicola Gallo.
// Conforms to the PIC Specification published and maintained by Nitro Agility S.r.l.

package scenario

import (
	"time"

	"github.com/pic-protocol/pic-prototyping/v0.2/golang/pic"
)

// This file runs the canonical Sandboxed Execution of the PIC Sandboxed
// Execution Specification: PIC carrying PIC. An AI agent holds the user's
// Lineage Execution A and mints its own Lineage Execution B; the two are
// proposed together as one Multi-Lineage Execution. An authorized sandbox origin
// originates the outer ENFORCE lineage (PCA0-G); the guardrail — an ordinary
// executor of that outer lineage — validates every carried lineage, evaluates
// the enforcement function, and on permit proves the next ordinary outer PCA
// carrying the signed multiLineage. A receiving hop then runs enforced acceptance.

// Grant identifiers bound to semantic scopes by the fixtures.
const (
	GrantUserBackup      = "urn:pic:grant:user-backup"
	GrantAgentS3Writer   = "urn:pic:grant:agent-s3-writer"
	GrantExternalSharing = "urn:pic:grant:external-sharing"
)

// Guardrail builds the fixture guardrail executor: its identity and attestation,
// the enforcement function (simulated PDP over the fixture policy), and the
// fixture scope bindings.
func (w *World) Guardrail() *pic.Guardrail {
	return pic.NewGuardrail(w.id("guardrail"), w.att("guardrail"), w.Set.Registry,
		&pic.LocalPDP{Policy: w.Set.Policy}, w.Set.Policy, w.Set.Scopes)
}

// EnforcementOrigin is the authorized sandbox origin that may mint PCA0-G.
func (w *World) EnforcementOrigin() *pic.Identity { return w.id("enforcement-origin") }

// AcceptedOrigins are the sandbox origins a receiving hop accepts.
func (w *World) AcceptedOrigins() []string { return []string{w.EnforcementOrigin().ID} }

// CrossingOutcome is one guarded crossing: the input Multi-Lineage Execution,
// the guardrail's enforcement trace, and (on permit) the outer ENFORCE lineage
// chain [PCA0-G, PCA1-G] that carries the decision.
type CrossingOutcome struct {
	Name       string                     `json:"name"`
	MLE        *pic.MultiLineageExecution `json:"multiLineageExecution"`
	Trace      *pic.EnforcementTrace      `json:"trace"`
	OuterChain []*pic.PCA                 `json:"outerChain,omitempty"`
	OuterPCA   *pic.PCA                   `json:"outerPca,omitempty"`
	Err        string                     `json:"error,omitempty"`
}

// ReceiverChecks is what a receiving hop does with the permitted crossing — and
// what happens to bypass and tamper attempts, under enforced acceptance.
type ReceiverChecks struct {
	Accepted       bool   `json:"accepted"`
	AcceptErr      string `json:"acceptError,omitempty"`
	BypassRejected bool   `json:"bypassRejected"`
	BypassReason   string `json:"bypassReason"`
	TamperRejected bool   `json:"tamperRejected"`
	TamperReason   string `json:"tamperReason"`
}

// GuardedResult is the full canonical Sandboxed Execution scenario.
type GuardedResult struct {
	Description string              `json:"description"`
	Origin      *pic.PCA            `json:"originPca0G"`
	Policy      pic.Policy          `json:"policy"`
	Scopes      map[string][]string `json:"scopeBindings"`
	Permit      *CrossingOutcome    `json:"permit"`
	Deny        *CrossingOutcome    `json:"deny"`
	InvalidPCA  *CrossingOutcome    `json:"invalidPca"`
	Receiver    ReceiverChecks      `json:"receiver"`
}

// Guarded runs the canonical example end to end. Every PCA and decision is
// really minted, signed, and verified on this call.
func (w *World) Guarded(now time.Time) (*GuardedResult, error) {
	g := w.Guardrail()
	agent := w.id("summary-service") // the AI agent (agentic fixture executor)
	agentAtt := w.att("summary-service")
	contract := pic.ExecutionContract{}

	// Lineage Execution A — user authority: alice grants {read-all, backup}; the
	// agent receives PCA1-A {backup} through a real PoR hop.
	pca0A, err := pic.MintPCA0(w.id("alice"),
		pic.Invariants{Operations: []string{"read-all", "backup"}, ExecutionContract: contract},
		GrantUserBackup, now)
	if err != nil {
		return nil, err
	}
	pca1A, err := pic.NewProver(agent, agentAtt).Continue(pca0A,
		pic.Invariants{Operations: []string{"backup"}, ExecutionContract: contract},
		pic.Request{Operation: "backup", Target: "/user/dataset", SecurityDomain: "tenant-42"}, now)
	if err != nil {
		return nil, err
	}
	chainA := []*pic.PCA{pca0A, pca1A}

	dest := "s3://backups/tenant-42"
	chainB, err := w.agentLineage(agent, agentAtt, GrantAgentS3Writer,
		[]string{"write:s3/backups/*"},
		pic.Request{Operation: "write", Target: "s3/backups/tenant-42/dataset.tar", SecurityDomain: "tenant-42"}, now)
	if err != nil {
		return nil, err
	}

	res := &GuardedResult{
		Description: "Canonical Sandboxed Execution (PIC of PIC): an authorized sandbox origin originates the outer ENFORCE lineage (PCA0-G). The AI agent holds the user's Lineage Execution A and its own Lineage Execution B and proposes the S3 write as one Multi-Lineage Execution. The guardrail — an ordinary executor of the outer lineage — validates every carried lineage, evaluates the enforcement function, and on permit proves the next ordinary outer PCA (PCA1-G) carrying the signed multiLineage. Authorities remain separate; nothing is merged.",
		Policy:      w.Set.Policy,
		Scopes:      w.Set.Scopes,
	}

	// Crossing 1 — PERMIT: A (data-protection) + B (data-protection, ai-compliance).
	sePermit, err := pic.Originate(w.EnforcementOrigin(), now)
	if err != nil {
		return nil, err
	}
	res.Origin = sePermit.Origin()
	mle := &pic.MultiLineageExecution{
		Participants: []pic.Participant{
			{Label: "A", Chain: chainA, Role: "user-backup"},
			{Label: "B", Chain: chainB, Role: "agent-s3-writer"},
		},
		Proposing:   "B",
		Destination: dest,
	}
	outer, trace, err := g.Enforce(sePermit, mle, now)
	res.Permit = &CrossingOutcome{Name: "A+B write to S3", MLE: mle, Trace: trace,
		OuterChain: sePermit.Chain, OuterPCA: outer, Err: errStr(err)}

	// Crossing 2 — DENY: A + C (external-sharing); the enforcement condition fails.
	chainC, err := w.agentLineage(agent, agentAtt, GrantExternalSharing,
		[]string{"share-public"},
		pic.Request{Operation: "share-public", Target: "s3/backups/tenant-42/dataset.tar", SecurityDomain: "tenant-42"}, now)
	if err != nil {
		return nil, err
	}
	seDeny, err := pic.Originate(w.EnforcementOrigin(), now)
	if err != nil {
		return nil, err
	}
	mleDeny := &pic.MultiLineageExecution{
		Participants: []pic.Participant{
			{Label: "A", Chain: chainA, Role: "user-backup"},
			{Label: "C", Chain: chainC, Role: "external-sharing"},
		},
		Proposing:   "C",
		Destination: "https://public.example/share",
	}
	outerDeny, traceDeny, err := g.Enforce(seDeny, mleDeny, now)
	res.Deny = &CrossingOutcome{Name: "A+C public share", MLE: mleDeny, Trace: traceDeny,
		OuterPCA: outerDeny, Err: errStr(err)}

	// Crossing 3 — INVALID carried lineage: a maliciously expanded successor B'.
	// The guardrail denies at validation, without evaluating policy.
	rogue, err := pic.NewProver(agent, agentAtt).ContinueMalicious(chainB[0],
		pic.Invariants{Operations: []string{"write:s3/backups/*", "delete:s3/*"}, ExecutionContract: contract},
		pic.Request{Operation: "delete", Target: "s3/backups/tenant-42", SecurityDomain: "tenant-42"}, now)
	if err != nil {
		return nil, err
	}
	seBad, err := pic.Originate(w.EnforcementOrigin(), now)
	if err != nil {
		return nil, err
	}
	mleBad := &pic.MultiLineageExecution{
		Participants: []pic.Participant{
			{Label: "A", Chain: chainA, Role: "user-backup"},
			{Label: "B'", Chain: []*pic.PCA{chainB[0], rogue}, Role: "agent-s3-writer"},
		},
		Proposing:   "B'",
		Destination: dest,
	}
	outerBad, traceBad, err := g.Enforce(seBad, mleBad, now)
	res.InvalidPCA = &CrossingOutcome{Name: "A+B' expanded authority", MLE: mleBad, Trace: traceBad,
		OuterPCA: outerBad, Err: errStr(err)}

	// Receiving hop: enforced acceptance of the permitted outer chain, plus
	// bypass (no outer chain) and tamper (edited carried lineage) rejections.
	res.Receiver = w.receiverChecks(res.Permit.OuterChain, now)
	return res, nil
}

// agentLineage mints the agent's own origin under the given grant and continues
// it once with the concrete signed request.
func (w *World) agentLineage(agent *pic.Identity, att pic.Attestation, grant string, ops []string, req pic.Request, now time.Time) ([]*pic.PCA, error) {
	contract := pic.ExecutionContract{}
	pca0, err := pic.MintPCA0(agent, pic.Invariants{Operations: ops, ExecutionContract: contract}, grant, now)
	if err != nil {
		return nil, err
	}
	pca1, err := pic.NewProver(agent, att).Continue(pca0,
		pic.Invariants{Operations: ops, ExecutionContract: contract}, req, now)
	if err != nil {
		return nil, err
	}
	return []*pic.PCA{pca0, pca1}, nil
}

// receiverChecks runs enforced acceptance on a permitted outer chain: accept it,
// reject a bypass (no outer chain), reject a tampered copy (edited carried lineage).
func (w *World) receiverChecks(outerChain []*pic.PCA, now time.Time) ReceiverChecks {
	rc := ReceiverChecks{}
	origins := w.AcceptedOrigins()

	verr := pic.AcceptGuardedCrossing(w.Set.Registry, nil, origins, outerChain, now)
	rc.Accepted = len(outerChain) > 0 && verr == nil
	rc.AcceptErr = errStr(verr)

	berr := pic.AcceptGuardedCrossing(w.Set.Registry, nil, origins, nil, now)
	rc.BypassRejected = berr != nil
	rc.BypassReason = errStr(berr)

	if len(outerChain) > 0 {
		tampered := tamperOuterChain(outerChain)
		terr := pic.AcceptGuardedCrossing(w.Set.Registry, nil, origins, tampered, now)
		rc.TamperRejected = terr != nil
		rc.TamperReason = errStr(terr)
	}
	return rc
}

// tamperOuterChain returns a copy of the outer chain whose tip carries a
// mutated multiLineage (edited destination), so the recomputed multiLineageDigest
// no longer matches the signed request (and the PCA signature no longer verifies).
func tamperOuterChain(outerChain []*pic.PCA) []*pic.PCA {
	out := make([]*pic.PCA, len(outerChain))
	copy(out, outerChain)
	tip := *out[len(out)-1]
	if tip.MultiLineage != nil {
		ml := *tip.MultiLineage
		ctx := ml.Context
		ctx.Destination = "s3://attacker/exfil"
		ml.Context = ctx
		tip.MultiLineage = &ml
	}
	out[len(out)-1] = &tip
	return out
}

// GuardTip carries an existing chain as Lineage Execution A through a Sandboxed
// Execution: the AI agent mints its own Lineage Execution B for the destination
// write and proposes the joint crossing. It is the shared `--guardrail`
// augmentation used by the other scenarios.
func (w *World) GuardTip(chain []*pic.PCA, destination string, now time.Time) (*CrossingOutcome, ReceiverChecks, error) {
	g := w.Guardrail()
	agent := w.id("summary-service")
	chainB, err := w.agentLineage(agent, w.att("summary-service"), GrantAgentS3Writer,
		[]string{"write:s3/backups/*"},
		pic.Request{Operation: "write", Target: "s3/backups/tenant-42/result.tar", SecurityDomain: "tenant-42"}, now)
	if err != nil {
		return nil, ReceiverChecks{}, err
	}
	se, err := pic.Originate(w.EnforcementOrigin(), now)
	if err != nil {
		return nil, ReceiverChecks{}, err
	}
	mle := &pic.MultiLineageExecution{
		Participants: []pic.Participant{
			{Label: "A", Chain: chain, Role: "user-backup"},
			{Label: "B", Chain: chainB, Role: "agent-s3-writer"},
		},
		Proposing:   "B",
		Destination: destination,
	}
	outer, trace, err := g.Enforce(se, mle, now)
	out := &CrossingOutcome{Name: "tip crossing", MLE: mle, Trace: trace,
		OuterChain: se.Chain, OuterPCA: outer, Err: errStr(err)}
	if err != nil {
		return out, ReceiverChecks{}, nil
	}
	return out, w.receiverChecks(se.Chain, now), nil
}

File: v0.2/rust/src/scenario/guardrail.rs
// SPDX-License-Identifier: Apache-2.0
//
// Based on the Provenance Identity Continuity (PIC) Model created by Nicola Gallo.
// Conforms to the PIC Specification published and maintained by Nitro Agility S.r.l.

//! The canonical Sandboxed Execution of the PIC Sandboxed Execution Specification
//! (PIC of PIC): an authorized sandbox origin originates the outer ENFORCE
//! lineage; the AI agent holds the user's Lineage Execution A and its own
//! Lineage Execution B, proposed together as one Multi-Lineage Execution; the
//! guardrail — an ordinary executor of the outer lineage — validates, evaluates,
//! and on permit proves the next outer PCA. Mirror of Go `scenario/guardrail.go`.

use super::World;
use crate::sandboxed::{
    accept_guarded_crossing, EnforcementTrace, Guardrail, LocalPdp, MultiLineageExecution,
    Participant, SandboxedExecution,
};
use crate::types::{Attestation, ExecutionContract, Invariants, Pca, Request};
use crate::{mint_pca0, Identity, PicResult, Prover};
use chrono::{DateTime, Utc};
use serde::Serialize;
use std::collections::HashMap;

/// Grant identifiers bound to semantic scopes by the fixtures.
pub const GRANT_USER_BACKUP: &str = "urn:pic:grant:user-backup";
pub const GRANT_AGENT_S3_WRITER: &str = "urn:pic:grant:agent-s3-writer";
pub const GRANT_EXTERNAL_SHARING: &str = "urn:pic:grant:external-sharing";

/// One guarded crossing: the input Multi-Lineage Execution, the guardrail's
/// enforcement trace, and (on permit) the outer ENFORCE lineage chain.
#[derive(Serialize)]
pub struct CrossingOutcome {
    pub name: String,
    #[serde(rename = "multiLineageExecution")]
    pub mle: MultiLineageExecution,
    pub trace: EnforcementTrace,
    #[serde(rename = "outerChain", skip_serializing_if = "Vec::is_empty")]
    pub outer_chain: Vec<Pca>,
    #[serde(rename = "outerPca", skip_serializing_if = "Option::is_none")]
    pub outer_pca: Option<Pca>,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub error: String,
}

/// What a receiving hop does with the permitted crossing, under enforced
/// acceptance.
#[derive(Default, Serialize)]
pub struct ReceiverChecks {
    pub accepted: bool,
    #[serde(rename = "acceptError", skip_serializing_if = "String::is_empty")]
    pub accept_err: String,
    #[serde(rename = "bypassRejected")]
    pub bypass_rejected: bool,
    #[serde(rename = "bypassReason")]
    pub bypass_reason: String,
    #[serde(rename = "tamperRejected")]
    pub tamper_rejected: bool,
    #[serde(rename = "tamperReason")]
    pub tamper_reason: String,
}

/// The full canonical Sandboxed Execution scenario.
#[derive(Serialize)]
pub struct GuardedResult {
    pub description: String,
    #[serde(rename = "originPca0G")]
    pub origin: Pca,
    pub policy: crate::Policy,
    #[serde(rename = "scopeBindings")]
    pub scopes: HashMap<String, Vec<String>>,
    pub permit: CrossingOutcome,
    pub deny: CrossingOutcome,
    #[serde(rename = "invalidPca")]
    pub invalid_pca: CrossingOutcome,
    pub receiver: ReceiverChecks,
}

impl World {
    /// The authorized sandbox origins a receiving hop accepts.
    pub fn accepted_origins(&self) -> Vec<String> {
        vec![self.id("enforcement-origin").id.clone()]
    }

    /// Runs the canonical example end to end.
    pub fn guarded(&self, now: DateTime<Utc>) -> PicResult<GuardedResult> {
        let pdp = LocalPdp {
            policy: self.set.policy.clone(),
        };
        let guardrail = Guardrail::new(
            self.id("guardrail"),
            self.att("guardrail"),
            &self.set.registry,
            &pdp,
            self.set.policy.clone(),
            &self.set.scopes,
        );
        let origin = self.id("enforcement-origin");
        let agent = self.id("summary-service");
        let agent_att = self.att("summary-service");
        let contract = ExecutionContract::default();

        // Carried lineage A — user authority.
        let pca0_a = mint_pca0(
            self.id("alice"),
            Invariants {
                operations: vec!["read-all".into(), "backup".into()],
                execution_contract: contract.clone(),
            },
            GRANT_USER_BACKUP,
            now,
        );
        let pca1_a = Prover::new(agent, agent_att.clone()).continue_(
            &pca0_a,
            Invariants {
                operations: vec!["backup".into()],
                execution_contract: contract.clone(),
            },
            Request {
                operation: "backup".into(),
                target: "/user/dataset".into(),
                security_domain: "tenant-42".into(),
                ..Default::default()
            },
            now,
        )?;
        let chain_a = vec![pca0_a, pca1_a];

        let dest = "s3://backups/tenant-42";
        let chain_b = self.agent_lineage(
            agent,
            &agent_att,
            GRANT_AGENT_S3_WRITER,
            &["write:s3/backups/*"],
            Request {
                operation: "write".into(),
                target: "s3/backups/tenant-42/dataset.tar".into(),
                security_domain: "tenant-42".into(),
                ..Default::default()
            },
            now,
        )?;

        // Crossing 1 — PERMIT.
        let mut se_permit = SandboxedExecution::originate(origin, now);
        let origin_pca = se_permit.origin().clone();
        let mle = MultiLineageExecution {
            participants: vec![
                participant("A", chain_a.clone(), "user-backup"),
                participant("B", chain_b.clone(), "agent-s3-writer"),
            ],
            proposing: "B".into(),
            destination: dest.into(),
        };
        let (outer, trace) = guardrail.enforce(&mut se_permit, &mle, now);
        let permit = CrossingOutcome {
            name: "A+B write to S3".into(),
            mle,
            error: err_of(&trace, outer.is_some()),
            outer_chain: se_permit.chain.clone(),
            outer_pca: outer,
            trace,
        };

        // Crossing 2 — DENY: A + C (external-sharing).
        let chain_c = self.agent_lineage(
            agent,
            &agent_att,
            GRANT_EXTERNAL_SHARING,
            &["share-public"],
            Request {
                operation: "share-public".into(),
                target: "s3/backups/tenant-42/dataset.tar".into(),
                security_domain: "tenant-42".into(),
                ..Default::default()
            },
            now,
        )?;
        let mut se_deny = SandboxedExecution::originate(origin, now);
        let mle_deny = MultiLineageExecution {
            participants: vec![
                participant("A", chain_a.clone(), "user-backup"),
                participant("C", chain_c, "external-sharing"),
            ],
            proposing: "C".into(),
            destination: "https://public.example/share".into(),
        };
        let (outer_deny, trace_deny) = guardrail.enforce(&mut se_deny, &mle_deny, now);
        let deny = CrossingOutcome {
            name: "A+C public share".into(),
            mle: mle_deny,
            error: err_of(&trace_deny, outer_deny.is_some()),
            outer_chain: Vec::new(),
            outer_pca: outer_deny,
            trace: trace_deny,
        };

        // Crossing 3 — INVALID carried lineage: a maliciously expanded B'.
        let rogue = Prover::new(agent, agent_att.clone()).continue_malicious(
            &chain_b[0],
            Invariants {
                operations: vec!["write:s3/backups/*".into(), "delete:s3/*".into()],
                execution_contract: contract.clone(),
            },
            Request {
                operation: "delete".into(),
                target: "s3/backups/tenant-42".into(),
                security_domain: "tenant-42".into(),
                ..Default::default()
            },
            now,
        )?;
        let mut se_bad = SandboxedExecution::originate(origin, now);
        let mle_bad = MultiLineageExecution {
            participants: vec![
                participant("A", chain_a, "user-backup"),
                participant("B'", vec![chain_b[0].clone(), rogue], "agent-s3-writer"),
            ],
            proposing: "B'".into(),
            destination: dest.into(),
        };
        let (outer_bad, trace_bad) = guardrail.enforce(&mut se_bad, &mle_bad, now);
        let invalid_pca = CrossingOutcome {
            name: "A+B' expanded authority".into(),
            mle: mle_bad,
            error: err_of(&trace_bad, outer_bad.is_some()),
            outer_chain: Vec::new(),
            outer_pca: outer_bad,
            trace: trace_bad,
        };

        let receiver = self.receiver_checks(&permit.outer_chain, now);
        Ok(GuardedResult {
            description: "Canonical Sandboxed Execution (PIC of PIC): an authorized sandbox origin originates the outer ENFORCE lineage (PCA0-G). The AI agent holds the user's Lineage Execution A and its own Lineage Execution B and proposes the S3 write as one Multi-Lineage Execution. The guardrail — an ordinary executor of the outer lineage — validates every carried lineage, evaluates the enforcement function, and on permit proves the next ordinary outer PCA (PCA1-G) carrying the signed multiLineage. Authorities remain separate; nothing is merged.".into(),
            origin: origin_pca,
            policy: self.set.policy.clone(),
            scopes: self.set.scopes.clone(),
            permit,
            deny,
            invalid_pca,
            receiver,
        })
    }

    /// Mints the agent's own origin under the given grant and continues it once
    /// with the concrete signed request.
    fn agent_lineage(
        &self,
        agent: &'static Identity,
        att: &Attestation,
        grant: &str,
        ops: &[&str],
        req: Request,
        now: DateTime<Utc>,
    ) -> PicResult<Vec<Pca>> {
        let inv = Invariants {
            operations: ops.iter().map(|s| s.to_string()).collect(),
            execution_contract: ExecutionContract::default(),
        };
        let pca0 = mint_pca0(agent, inv.clone(), grant, now);
        let pca1 = Prover::new(agent, att.clone()).continue_(&pca0, inv, req, now)?;
        Ok(vec![pca0, pca1])
    }

    /// Enforced acceptance of a permitted outer chain, plus bypass and tamper
    /// rejections.
    pub fn receiver_checks(&self, outer_chain: &[Pca], now: DateTime<Utc>) -> ReceiverChecks {
        let mut rc = ReceiverChecks::default();
        let origins = self.accepted_origins();

        let verr = accept_guarded_crossing(&self.set.registry, None, &origins, outer_chain, now);
        rc.accepted = !outer_chain.is_empty() && verr.is_ok();
        rc.accept_err = verr.err().unwrap_or_default();

        let berr = accept_guarded_crossing(&self.set.registry, None, &origins, &[], now);
        rc.bypass_rejected = berr.is_err();
        rc.bypass_reason = berr.err().unwrap_or_default();

        if !outer_chain.is_empty() {
            let mut tampered = outer_chain.to_vec();
            if let Some(tip) = tampered.last_mut() {
                if let Some(ml) = tip.multi_lineage.as_mut() {
                    ml.context.destination = "s3://attacker/exfil".into();
                }
            }
            let terr = accept_guarded_crossing(&self.set.registry, None, &origins, &tampered, now);
            rc.tamper_rejected = terr.is_err();
            rc.tamper_reason = terr.err().unwrap_or_default();
        }
        rc
    }

    /// Carries an existing chain as carried lineage A through a Sandboxed
    /// Execution (the shared `--guardrail` augmentation).
    pub fn guard_tip(
        &self,
        chain: Vec<Pca>,
        destination: &str,
        now: DateTime<Utc>,
    ) -> PicResult<(CrossingOutcome, ReceiverChecks)> {
        let pdp = LocalPdp {
            policy: self.set.policy.clone(),
        };
        let guardrail = Guardrail::new(
            self.id("guardrail"),
            self.att("guardrail"),
            &self.set.registry,
            &pdp,
            self.set.policy.clone(),
            &self.set.scopes,
        );
        let agent = self.id("summary-service");
        let chain_b = self.agent_lineage(
            agent,
            &self.att("summary-service"),
            GRANT_AGENT_S3_WRITER,
            &["write:s3/backups/*"],
            Request {
                operation: "write".into(),
                target: "s3/backups/tenant-42/result.tar".into(),
                security_domain: "tenant-42".into(),
                ..Default::default()
            },
            now,
        )?;
        let mut se = SandboxedExecution::originate(self.id("enforcement-origin"), now);
        let mle = MultiLineageExecution {
            participants: vec![
                participant("A", chain, "user-backup"),
                participant("B", chain_b, "agent-s3-writer"),
            ],
            proposing: "B".into(),
            destination: destination.into(),
        };
        let (outer, trace) = guardrail.enforce(&mut se, &mle, now);
        let rc = if outer.is_some() {
            self.receiver_checks(&se.chain, now)
        } else {
            ReceiverChecks::default()
        };
        let out = CrossingOutcome {
            name: "tip crossing".into(),
            mle,
            error: err_of(&trace, outer.is_some()),
            outer_chain: se.chain.clone(),
            outer_pca: outer,
            trace,
        };
        Ok((out, rc))
    }
}

fn participant(label: &str, chain: Vec<Pca>, role: &str) -> Participant {
    Participant {
        label: label.into(),
        chain,
        role: role.into(),
    }
}

/// The error string of a denied crossing (empty on permit), mirroring Go.
fn err_of(trace: &EnforcementTrace, permitted: bool) -> String {
    if permitted {
        String::new()
    } else {
        format!("guardrail: deny — {}", trace.decision.reason)
    }
}
